### PR TITLE
Back button handler

### PR DESCRIPTION
**Pull request by lcaprini**
Hi NoriginMedia,
me and my team are using your library in our react TV app and we made a small improvement in order to handle the back button.

The Back event could be handled defining the new `onBackPress` callback in the `useFocusable` configuration: this callback could be used in both component and Focusable Context.

As README.md explains, that callback is called when the component is focused and Back key is pressed. If no callback was provided to the focused component, the event will go up through the Focusable Tree as long as one defined callback was found or the top was reached up.

The `back` key is set to `27`, like the ESC key; of course it could be set with `setKeyMap` method.

A small example of the behaviour was provided on `App.tsx` demo page: when the back event was fired while the focus is on a `Content`, the focus jumps to the side menu; if the back event was fired while the focus is on the "Asset 3" element, an alert will show up.

I hope it could be usefull.

--Luca

### CODE REVIEW

**Review by emilehay — APPROVED** [review]
I've tested this change and it works as expected, adding much-needed functionality to the library.

As for the implementation, the `onBackPress` property mirrors the `onEnterPress` functionality almost 1:1 just mapped to a different key, so I don't foresee any unexpected regressions from this change.

**Comment by siarheipashkevich** [comment]
I also need this feature )

**Comment by siarheipashkevich** [comment]
@lcaprini @emilehay can we extend this logic to handle any button handlers? For example `play` `pause` etc.

**Comment by lcaprini** [comment]
> @lcaprini @emilehay can we extend this logic to handle any button handlers? For example `play` `pause` etc.

Of course is possible, but in my opinion the play, pause and other remote buttons aren't directly related to the navigation. I do it in a different library/package.

**Comment by siarheipashkevich** [comment]
@lcaprini but before handling key events we should check that the needed element has focus and etc..

Could you please share package?

**Comment by lcaprini** [comment]
> but before handling key events we should check that the needed element has focus and etc..

My package listens the keydown event without consider the focused element, because the playback remote buttons are used across the whole app.

My package is currently on a private repo; I'll publish also in github asap.

**Comment by siarheipashkevich** [comment]
@lcaprini can you share your package please?

**Comment by lcaprini** [comment]
> @lcaprini can you share your package please?

The first draft version is online on npm as [react-remote-controller](https://www.npmjs.com/package/react-remote-controller). There is no doc on it actually, but you could install with `npm i react-remote-controller` and see the example in `src/example/App.tsx`. I write the README.md when I have time.

**Comment by siarheipashkevich** [comment]
@lcaprini thank you so much. I'll investigate your package.

**Comment by diolkaee** [comment]
We are also gratefully using your library & eagerly awaiting this feature. Can you foresee when this change will be integrated?

**Comment by guilleccc** [comment]
Hi,

I have been thinking about this feature, and honestly, I don't see any good reason to add this inside Norigin-Spatial-Navigation. Since the goal of this library is to handle spatial navigation (directional navigation).
And I think @siarheipashkevich had a good point, if we add back logic, why not other keys as well?

Then in my personal opinion separating the code into two parts will be the best:
```
  const { ref, focused } = useFocusable();
  const {backHandler} = useBack(); // or useKeyListener
```
Thus the app developer will implement its own `useBack` on the app code.

**Comment by lcaprini** [comment]
Hi @guillecc, thank you for your reply. Your considerations about the back buttons could be correct, but them could be also applyed to the enter button as well.
Also the enter button is not strictly connected to the navigation, but to the action related to the focused element. In the same way, the back button action is related to a focusable item or context.

**Comment by drudolf** [comment]
Hi,

as I needed more keys to work in the context of the focus tree I made a more broader approach, introducing a `onKeyPress` handler, which could also take care of `back`. I needed this for example to be able to handle color keys in different parts of our app too. Also needed to be able to use arrow keys for some functionality when there is nothing to navigate to like closing a modal.

Would love to hear you feedback on it, https://github.com/drudolf/Norigin-Spatial-Navigation/tree/feature/key-press.

Thanks.

**Comment by YoavHortman** [comment]
Based on having onEnterKey, onBackKey imo is a natural extension.
I think adding other keys is not in the spirit of this library - it is a navigation tool, and a classic property of navigational tools is to go back.

**Comment by asgvard** [comment]
Hello.

We had a lot of internal discussions on this and the consensus was that we are not really keen on adding Back handler to this library.
More explanations here: https://github.com/NoriginMedia/Norigin-Spatial-Navigation/issues/57

TLDR; `onEnterPress` is mostly used to listen to the currently focused element being pressed, while `onBackPress` is mostly used to navigate to the previous page, so it's not usually linked to the currently focused element. Of course there are exceptions, like "if player controls are open and Back is pressed, close the controls", but this is easily achievable with a simple callback on Back button key press where you check whether "player controls are open" or similar.
We are not sure that we should go with the rule "since we have X button, let's add another A, B, C key listeners to the library". It will bloat the library with the logic that is not directly related to the spatial (directional) navigation, and all the extra buttons listeners can be added in a custom way in the project. Same goes with the gamepads support and/or other devices with different key codes. We'd like to keep this library device-agnostic. We might even remove the default key listener since now it assumes the default browser key codes, so we might make it pluggable and provide an exportable default implementation.
